cunumeric/module.py

```diff
@@ -1330,6 +1332,131 @@ def append(arr, values, axis=None):
+        If passed a single ndarray or scalar (a nested list of depth 0),
+        this is returned unmodified (and not copied).
+
+        Elements shapes must match along the appropriate axes (without
+        broadcasting), but leading 1s will be prepended to the shape as
+        necessary to make the dimensions match.
+
+    Returns
+    -------
+    block_array : ndarray
+        The array assembled from the given blocks.
+        The dimensionality of the output is equal to the greatest of: * the
+        dimensionality of all the inputs * the depth to which the input list
+        is nested
+
+    Raises
+    ------
+    ValueError
+        If list depths are mismatched - for instance, [[a, b], c] is
+        illegal, and should be spelt [[a, b], [c]]
+        If lists are empty - for instance, [[a, b], []]
+
+    See Also
+    --------
+    numpy.block
+
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+
+    """
+    # arrays should concatenate from innermost subarrays
+    # the 'arrays' should be balanced tree
+    active_arr = deque()
+    active_arr.append(((0,), arrays))
+    leaf_map = defaultdict(list)
+
+    # check if the 'arrays' is a balanced tree
+    while len(active_arr) > 0:
+        top = active_arr.popleft()
+        if type(top[1]) is list:
+            # check if this is a leaf
+            nlist = 0
+            for arr in top[1]:
+                if type(arr) is list:
+                    active_arr.append(((*top[0], nlist), arr))
+                    nlist += 1
+            # this is a leaf
+            if nlist == 0:
+                if len(top[1]) == 0:
+                    ValueError(f"List at {top[0]} " f"cannot be empty")
+                leaf_map[len(top[0])].append((top[0], top[1]))
+            elif nlist < len(top[1]):
+                ValueError(
+                    f"some elements in depth {len(top[0])} "
+                    f"have different depths"
+                )
+    keys = list(leaf_map.keys())
+    if len(keys) > 1:
+        ValueError(
+            f"List depths for innermost elements are not matched"
+            f"array[keys[0]] is at depth {len(keys[0])}"
+            f"array[keys[1]] is at depth {len(keys[1])}"
+        )
+    # fill the working deque w/ innermost arrays
+    active_arr.extend(*leaf_map.values())
+    depth = keys[0]
+
+    while len(active_arr) > 0 and len(active_arr[-1][0]) > 0:
+        elem = active_arr.pop()
+        cur_depth = len(elem[0])
+        siblings = []
+        max_ndim = 0
+        # collect siblings under the same parent
+        while len(active_arr) > 0 and active_arr[-1][0] == elem[0]:
+            siblings.append(active_arr.pop()[1])
+
+        if len(siblings) > 0 or cur_depth < depth:
+            siblings.append(elem[1])
+        elif cur_depth == depth:
+            siblings = elem[1]
+
+        inputs = list(
+            ndarray.convert_to_cunumeric_ndarray(inp) for inp in siblings
```

Comment:
do we need to do this in every iteration? I think once the inputs are sanitized, everything should be a cunumeric ndarray.

Reply:
during the first `while` loop, it just checks the depth of lists. The innermost arrays may contain scalars. We can restrict this to happen only for the concatenation at the leaf level. I'll do it.